[FEATURE] Microsoft teams notifications

Antoninj opened the pull request:
Changes proposed in this pull request:

- Add a MicrosoftTeamsNotificationAction

Antoninj committed:
Implement `send_microsoft_teams_notifications` and more generic `send_webhook_notifications` utility methods

Antoninj committed:
Implement MicrosoftTeamsNotificationAction

Antoninj committed:
Implement MicrosoftTeamsRenderer

Antoninj committed:
Update changelog.rst

Antoninj committed:
Add MicrosoftTeamsNotificationAction to ActionAnonymizer

Antoninj committed:
Import MicrosoftTeamsNotificationAction in validation operators module

Antoninj committed:
Update MicrosoftTeamsRenderer implementation

Antoninj committed:
Fix microsoft teams renderer

Antoninj committed:
Implement basic unit test of MicrosoftTeamsNotificationAction for success and failure notifications use cases

Antoninj committed:
Pass validation_result_suite_identifier to MicrosoftTeamsRenderer

Antoninj committed:
Add validation_result_suite_extended_id test fixture

Antoninj committed:
Update MicrosoftTeamsRenderer implementation to make use of validation_result_suite_identifier

Antoninj commented:
@cla-bot check

eugmandel committed:
Merge branch 'develop' into feature/microsoft-teams-notifications

eugmandel commented on tests/actions/test_core_actions.py:
```diff
@@ -135,6 +137,45 @@ def test_PagerdutyAlertAction(
+        "module_name": "great_expectations.render.renderer.microsoft_teams_renderer",
+        "class_name": "MicrosoftTeamsRenderer",
+    }
+    teams_webhook = "https://webhook.site/700303fe-6398-4a22-96fb-837f90d019e9"
```
Is this a special test token?

Antoninj replied:
It is an expiring url unfortunately:

- http://docs.webhook.site/premium.html#unlimited-requests
- http://docs.webhook.site/premium.html#non-expiring-urls

 We need to find a better way to deal wit this to run this test.

Shinnnyshinshin replied:
@Antoninj one pattern you may be able to follow is the test for `PagerdutyAlertAction` (`test_PagerdutyAlertAction()` is in the same `test_core_actions.py` file)

For `PagerdutyAlertAction`, the `_run` function's response does not come from the actual the POST, but is `success` or `none sent` depending on whether the POST event sent as expected.   

Admittedly, this is not a 'full' test, but in combination with the `test_MicrosoftTeams_validation_results_with_datadocs()` test you wrote, I think we can be 'reasonably' confident that notification will behave as expected.

eugmandel reviewed (changes requested):
@Antoninj This is a great addition! I went over the code and it makes sense to me. One request: I don't use MS teams, so it is not easy for me to see the notification. Could you record a video (Loom or anything else) of configuring and using the feature? This video will be also useful for putting in the documentation and into promoting this feature. 

Thank you!

eugmandel committed:
Merge branch 'develop' into feature/microsoft-teams-notifications

Antoninj committed:
Add microsoft teams renderer unit test

Antoninj committed:
Merge branch 'develop' into feature/microsoft-teams-notifications

# Conflicts:
#	docs/changelog.rst

Antoninj committed:
Update changelog.rst

Antoninj committed:
Fix microsoft teams renderer unit test (remove print statements)

Antoninj committed:
Fix microsoft teams renderer unit test

Antoninj committed:
Upddate webhook site url because of this:

 - http://docs.webhook.site/premium.html#unlimited-requests
 - http://docs.webhook.site/premium.html#non-expiring-urls

 Need to find a better way to deal wit this to run this test

Shinnnyshinshin commented:
> @Antoninj This is a great addition! I went over the code and it makes sense to me. One request: I don't use MS teams, so it is not easy for me to see the notification. Could you record a video (Loom or anything else) of configuring and using the feature? This video will be also useful for putting in the documentation and into promoting this feature.
> 
> Thank you!

I also wanted to echo @eugmandel words in thanking you for this addition, and a small request for a video for configuring and using the feature. Overall this looks wonderful :)

Antoninj commented:
Thanks for the positive feedback @Shinnnyshinshin. I recorded a loom video that demoes the feature and shared it with @eugmandel a few weeks ago.

Shinnnyshinshin commented:
Hi @Antoninj thanks for putting that loom video together. It'll be a great resource for us when writing the docs and promotion :) 

Regarding the remaining changes, is there anything we can do from our end that will make things smoother for you?

Shinnnyshinshin commented:
Hi @Antoninj I wanted to do a short follow up on this PR with you.

First, I wanted to say that we are very excited to get this feature out. This is some fantastic work, and you've basically finished 95% of it already :) 

We'd like to offer some help with the remaining 5%. If by 1/4/21 you haven't had a chance to follow up, then we'd be happy to implement the remaining changes to the tests and merge it with the GE codebase as a new feature (with you getting credit for the work of course)

If you'd like to do the remaining 5% before then, then please let us know and we'll hold off on making any additions or changes to this PR. 

Either way, thank you very very much.  Contributors like yourself are what make this project possible.

Antoninj commented:
Hello @Shinnnyshinshin, 

Thanks again for the nice words. I really enjoyed the experience of contributing to the project so far and would love to complete the PR. I've been very busy in the previous weeks but I definitely think I'll be able to push the remaining changes by the 4th. I might get in touch with you on slack to clarify a few things.

Best wishes, 

Antonin

Antoninj committed:
Merge branch 'develop' into feature/microsoft-teams-notifications

Antoninj committed:
Fix bug that arises after incorrect merge conflicts fix

Shinnnyshinshin commented:
thank you @Antoninj, we really appreciate you taking the time. I'll follow up in a few days :)

Shinnnyshinshin commented:
Hi @Antoninj checking in. I'd be happy to get this PR over the line before EOD. Please let me know if you'd like me to hold off

Antoninj commented:
@Shinnnyshinshin feel free to finish it, I'm unfortunately not in a position to spend more time on this atm.

Shinnnyshinshin commented:
changes have been merged with #2246 :) Thank you again @Antoninj . looking forward to working with you again.
